streams/test-utils/src/main/java/org/apache/kafka/streams/processor/MockProcessorContext.java

```diff
@@ -0,0 +1,468 @@
+        //noinspection unchecked
+        capturedForwards.add(new CapturedForward(to.childName, new KeyValue(key, value)));
+    }
+
+    @Override
+    public <FK, FV> void forward(final FK key, final FV value, final int childIndex) {
+        //noinspection unchecked
+        capturedForwards.add(new CapturedForward(childIndex, new KeyValue(key, value)));
+    }
+
+    @Override
+    public <FK, FV> void forward(final FK key, final FV value, final String childName) {
+        //noinspection unchecked
+        capturedForwards.add(new CapturedForward(childName, new KeyValue(key, value)));
+    }
+
+    /**
+     * A method for retrieving all the forwarded data this context has observed. The returned list will not be
+     * affected by subsequent interactions with the context. The data in the list is in the same order as the calls to
+     * {@code forward(...)}.
+     *
+     * @return A list of key/value pairs that were previously passed to the context.
+     */
+    public List<KeyValue> forwarded() {
+        final LinkedList<KeyValue> result = new LinkedList<>();
+        for (final CapturedForward capturedForward : capturedForwards) {
+            result.add(capturedForward.kv);
+        }
+        return result;
+    }
+
+    /**
+     * A method for retrieving all the forwarded data this context has observed for a specific child by index.
+     * The returned list will not be affected by subsequent interactions with the context.
+     * The data in the list is in the same order as the calls to {@code forward(...)}.
+     *
+     * @param childIndex The child index to retrieve forwards for
+     * @return A list of key/value pairs that were previously passed to the context.
+     * @deprecated please re-write processors to use {@link #forward(Object, Object, To)} instead
+     */
+    @Deprecated
+    public List<KeyValue> forwarded(final int childIndex) {
+        final LinkedList<KeyValue> result = new LinkedList<>();
+        for (final CapturedForward capture : capturedForwards) {
+            if (capture.childIndex != null && capture.childIndex == childIndex) {
+                result.add(capture.kv);
+            }
+        }
+        return result;
+    }
+
+    /**
+     * A method for retrieving all the forwarded data this context has observed for a specific child by name.
+     * The returned list will not be affected by subsequent interactions with the context.
+     * The data in the list is in the same order as the calls to {@code forward(...)}.
+     *
+     * @param childName The child name to retrieve forwards for
+     * @return A list of key/value pairs that were previously passed to the context.
+     */
+    public List<KeyValue> forwarded(final String childName) {
+        final LinkedList<KeyValue> result = new LinkedList<>();
+        for (final CapturedForward capture : capturedForwards) {
+            if (capture.childName != null && capture.childName.equals(childName)) {
```

Comment:
Hmm.. is this correct? If `forward(kv)` is called without childName or childIndex, it means sending to all children. So should this be `capture.childName == null || ...` ? Ditto above in line 414.

Reply:
ah, yeah, that makes more sense. I was thinking in terms of reflecting exactly what calls to `forward(...)` there were, but it's probably more useful to mimic the regular forwarding logic.